Implement Backdrop

saecki opened the pull request:
This implements the material backdrop as an `Openable` so that it will behave just like a navigation drawer.
I also changed some things that I think make the interface feel more modern. If you don't like them I will undo them.
- Slightly round the corners of the artwork in the now playing tab to make it integrate better with the backdrop. 
- Highlight the currently playing song inside the queue by coloring the text and icon in primary colors instead of coloring the background.
- Remove the separators in between item of the collection overview.
- Change the color of the bottom navigation view. (In light mode in the pictures it looks barely distinguishable from the main content but in person there is a clear shadow. Probably due to the emulator)

### Open backdrop
The darkened area actually covers the whole front layout if the backdrop menu is opened but the emulator fails to render that.
Dark | Light
------ | -------
![image](https://user-images.githubusercontent.com/43008152/113170739-1fdf6700-9247-11eb-9fe9-5312d58e830e.png) | ![image](https://user-images.githubusercontent.com/43008152/113170774-28d03880-9247-11eb-87fd-18595607036d.png)
![image](https://user-images.githubusercontent.com/43008152/113170995-5fa64e80-9247-11eb-8a73-fa4beaaaae63.png) | ![image](https://user-images.githubusercontent.com/43008152/113171027-67fe8980-9247-11eb-9b26-c193771bdd57.png)

### Collection

Dark | Light
------ | -------
![image](https://user-images.githubusercontent.com/43008152/113167520-16a0cb00-9244-11eb-94f4-876edd328c1d.png) | ![image](https://user-images.githubusercontent.com/43008152/113168617-1ead3a80-9245-11eb-95f0-a2877a639dd6.png)

### Queue

Dark | Light
------ | -------
![image](https://user-images.githubusercontent.com/43008152/113168763-3edcf980-9245-11eb-9769-84fe3b329014.png) | ![image](https://user-images.githubusercontent.com/43008152/113168792-47cdcb00-9245-11eb-98c6-4e44518ab8f6.png)

### Now playing


Dark | Light
------ | -------
![image](https://user-images.githubusercontent.com/43008152/113168898-616f1280-9245-11eb-93a5-32a62034ec3d.png) | ![image](https://user-images.githubusercontent.com/43008152/113169042-82cffe80-9245-11eb-8fe1-6d5365338648.png)

saecki committed:
initial backdrop implementation

saecki committed:
use xml icons in the toolbar to fix coloring

saecki committed:
fix clipping for backdrop layout

saecki committed:
small theme changes

saecki committed:
cleanup finding backdroplayout

saecki committed:
frame player artwork inside a cardview

saecki committed:
remove dividers on the collection view

saecki committed:
theme changes and cleanup

saecki committed:
update kotlin version

agersant commented:
At first glance, this looks great!! Thanks a lot once again!
I'll try to QA the changes and read the diffs this weekend.

saecki committed:
initialize toolbar in onCreate

saecki committed:
cleanup

saecki commented:
It seems like if we use `setupActionBarWithNavController` to set up the toolbar the toolbar title doesn't say "Polaris" on startup and on screen rotation, but the backdrop is not openable. If we use `Toolbar.setupWithNavController` the backdrop is openable but the title will say Polaris. I'm looking for a solution right now.

saecki committed:
fix toolbar showing app name

saecki commented:
Well that turned out to be little change but it seems to be fixed for me.

Edit: Never mind still issues with the title in the now playing tab.

saecki committed:
fix toolbar not displaying current song title and artist

saecki commented:
Okay I really hope there aren't any more gotchas here.

agersant commented:
I couldn't get it to break with a few minutes of testing, and the diffs look good to me (although there's a lot of stuff I'm not super familiar with).

Removing the separators on the Collection screen was a great idea. Maybe we should also remove the greyed out `Playlists` button considering it's purely placeholder for now (and has been for years).

There's two things I think could be improved:
- The transitions between `Collection`/`Queue`/`NowPlaying` with the crossfade feel a bit slow
- When the settings are opened, the only way to close them is to click again on `Collection`/`Queue`/`NowPlaying`. However, one of them is still highlighted at this point so it's a bit confusing that it needs re-tapping to restore. Not sure if that's expected with the Backdrop pattern but it felt off.

saecki committed:
remove settings fragment from appbarconfig

saecki commented:
Ok when in the settings fragment the navigation icon will now show a back arrow. As far as the crossfade animations go I didn't change anything intentionally. They seem to be default for bottom navigation views. I barely even noticed it to be honest, is it just to slow in your opinion or do you want it completely removed.

agersant commented:
> Ok when in the settings fragment the navigation icon will now show a back arrow. 

Perfect

> I barely even noticed it to be honest, is just to slow in your opinion or do you want it completely removed.

I think I would remove them completely, but I could settle for faster if that's easier.

I did more testing last night and I noticed a more serious issue:
- Browse to a specific subdirectory within `Collection`, or to `Random`/`Recent`
- Tap the `Queue` or `Player` icon
- Press the back button,
- You end up on the collection root, instead of back where you were within the collection

saecki committed:
Improve navigation

- use separate nav controllers for bottom navigation tabs
- set content specific titles in the browse fragment
- remove crossfade transitions between root fragments

saecki commented:
I think the navigation behavior was introduced with the jetpack navigation. The original material design spec for the bottom navigation view intended that behavior. Since then google changed their minds but haven't added official support for separate backstacks (histories) in different tabs. In [this repository](https://github.com/android/architecture-components-samples/blob/master/NavigationAdvancedSample/app/src/main/java/com/example/android/navigationadvancedsample/NavigationExtensions.kt) I found an extensions that implements the wanted behavior and just copied it.
Turns out that it also allows changing transistions between fragments.
I also added specific titles on the browse fragment depending on which content (`Browse directories`, `Random albums` or `Recently added`) is selected.

saecki committed:
remove unused imports and variables

saecki committed:
fix repeated deeplink handling on screen rotation by navigation extension

saecki committed:
fix browse content specific toolbar titles overwritten on screen rotation

saecki committed:
cleanup backdrop menu and layout

saecki commented:
While in use I found some other regressions that should be fixed now.

saecki commented:
Rebased on top of master.

agersant commented on app/src/main/java/agersant/polaris/navigation/NavigationExtensions.kt:
```diff
@@ -0,0 +1,251 @@
+/*
```
Can you add a comment near the top describing where exactly this code came from?

saecki replied:
Done

agersant commented:
Changes look good (minus one trivial comment). During testing, I noticed more paper cuts related to the prior navigation changes, but I'm OK deferring these for a future PR:

- Returning to previously browsed screen triggers a new fetch of the data, which can end up using a lot of bandwidth and slowing down navigation
- Returning to a previously browsed screen doesn't restore scrolling position, which is inconvenient for very long directories

Thanks again for the hard work!

saecki committed:
add comment linking repository of NavigationExtensions

saecki commented:
> - Returning to previously browsed screen triggers a new fetch of the data, which can end up using a lot of bandwidth and slowing down navigation
> - Returning to a previously browsed screen doesn't restore scrolling position, which is inconvenient for very long directories

I've noticed that too, and already implemented a solution using viewmodels to cache data and store the scroll position on my [collection](https://github.com/Saecki/polaris-android/tree/collection/) branch. But Like you said that can wait for another PR.

Before that though I will probably create a PR for the [player](https://github.com/Saecki/polaris-android/tree/player/) branch, since the collection branch is based on that.

agersant commented:
Pushed a new Beta release to Google Play with these changes too.
